### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by fangpenlin**
Restart from a solid ground up

Basic models are built (core business logic and RESTful API are not implemented yet), still not in functional stage. However, we have 97% testing coverage currently. I hope I can keep this through and eventually deliver with 100% testing coverage.

```
Name                        Stmts   Miss  Cover   Missing
---------------------------------------------------------
billy                           0      0   100%
billy.models                   19      9    53%   19-36
billy.models.company           37      0   100%
billy.models.customer          36      0   100%
billy.models.plan              48      0   100%
billy.models.subscription      45      1    98%   89
billy.models.tables            81      1    99%   35
billy.models.transaction       42      0   100%
billy.utils                     0      0   100%
billy.utils.generic            23      0   100%
---------------------------------------------------------
TOTAL                         331     11    97%
----------------------------------------------------------------------
Ran 39 tests in 1.282s

OK
```

To install and run testing

```
virtualenv --no-site-packages env
source env/bin/activate
python setup.py develop
python setup.py nosetests
```

Only merge this branch back when it is good to go.

### CODE REVIEW

File: billy/models/subscription.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+
+class SubscriptionCanceledError(RuntimeError):
+    """This error indicates that the subscription is already canceled,
+    you cannot cancel a canceled subscription
+
+    """
+
+
+class SubscriptionModel(object):
```
Comment (mjallday) [inline billy/models/subscription.py:20]: @victorlin what's your thinking behind suffixing all the models with Model (e.g. `SubscriptionModel`) and then also including the type in the method name? 

This seems very verbose:

``` python
from billy.models import SubscriptionModel

subscription = SubscriptionModel.create_subscription(...)    
subscription = SubscriptionModel.get_subscription_by_guid(xxx)
subscription.update_subscription(...)
```

To me, calling the model `Subscription` and the methods `get` and `update` respectively seems more concise and it doesn't detract from readability.

File: billy/models/plan.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        external_id=None, 
+        name=None, 
+        description=None,
+    ):
+        """Create a plan and return its ID
+
+        """
+        if plan_type not in self.TYPE_ALL:
+            raise ValueError('Invalid plan_type {}'.format(plan_type))
+        if frequency not in self.FREQ_ALL:
```
Comment (pnegahdar) [inline billy/models/plan.py:72]: Why are we limiting ourselves to 4 intervals? One of the key features of billy (over the competition) was allowing arbitrary intervals...

File: billy/models/plan.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            frequency=frequency, 
+            external_id=external_id, 
+            name=name, 
+            description=description,
+        )
+        self.session.add(plan)
+        self.session.flush()
+        return plan.guid
+
+    def update_plan(self, guid, **kwargs):
```
Comment (pnegahdar) [inline billy/models/plan.py]: Why is plan guid being passed in? Is a instantiated model not self referential. Also this type of kwarg usage makes the model api hard to read and understand... I have yet to clone this to do a full review but unless its a classmethod the instantiated model classes should be aware of the object being updated...

File: billy/models/subscription.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        :param raise_error: Raise KeyError when cannot find one
+        """
+        query = self.session.query(tables.Subscription) \
+            .filter_by(guid=guid) \
+            .first()
+        if raise_error and query is None:
+            raise KeyError('No such subscription {}'.format(guid))
+        return query
+
+    def create_subscription(
```
Comment (pnegahdar) [inline billy/models/subscription.py]: Guids are being passed everywhere, pass objects.

Comment (pnegahdar) [inline billy/models/subscription.py]: Billy's rewrite was to eliminate these types of things all across the board. You're switching back to a CRUD based model as opposed to a fluid relationship based model

File: billy/models/subscription.py (hunk 1 above)
Comment (fangpenlin) [inline billy/models/subscription.py:20]: @mjallday I like to provide expressive naming in most cases, however, yeah, you're right, this is a overkill here. In certain model context, this cannot help readability afterall. I will correct them later.

But for Model suffixing, I think this is fine. The whole model here is different from what it used to be. I try to isolate business logic from SQLAlchemy object and operations. I don't want to expose too much SQLAlchemy details as when one day we like to switch NoSQL or something like that, it would be difficult to get rid of SQLAlchemy everywhere in the project, as I explained in #24. Also, there is table class named `Subscription`, I don't want them to be confusing. And from the big picture, it would be a MVC structure, the naming might look like this

```
[Subscription (table)] <---- [SubscriptionModel (business logic)] <---- [API web controller]
```

So, I will keep Model suffixing. Until I make this system workable, if you don't like naming style, I can make some change here.

File: billy/models/subscription.py (hunk 4 above)
Comment (fangpenlin) [inline billy/models/subscription.py]: I got reasons to pass only GUID. Passing SQLALchemy table class instance around is error-prone in certain situations. For example, when you are dealing with multi-threading, passing object instance cross threads causes problem. 

``` python
def process(records):
    for record in records:
        with transaction.manager:
            record.num += 1
            session.add(record)

pool = threading_pool()
for thread in pool:
    records = get_records(100)
    thread.run(process, records)
```

Let's say, we are processing transactions and we like to speed it up by using multi-threading or multi-processor. The code we wrote above looks fine, but it is not. You cannot create SQLALchemy object in one thread and pass it to another (exception will be raised). There is some thread-local stuff Session object is dealing with (sessions are bound to local threads). So, in this case, I prefer to make model only accept GUID argument rather than object. 

There are some other reasons, but I cannot recall (like dealing with database transaction or what).

File: billy/models/plan.py (hunk 3 above)
Comment (fangpenlin) [inline billy/models/plan.py]: The way I design it to be used is like

``` python
model = FoobarModel(session)
model.do_something()
```

Creating a model is very cheap, there is noting it creates internally, that's just all about business logic. So you can create them around everywhere, use them and drop them as you like. 

You may ask why not just use the session in a global variable or class variable manner. Well, because "global" means dirty in many cases. For example, when we want to use multi-threading, global variable makes it difficult to rewrite the code. You will need a session for each thread. Sometimes you may need to connect to multiple databases if you are doing sharding in database, a global session makes it hard to do this kind of things. 

In short, for many cases, dependency to a global session object makes it hard to do many important improvements afterward. This is why I like to pass session object only when it is needed.

File: billy/models/plan.py (hunk 2 above)
Comment (fangpenlin) [inline billy/models/plan.py:72]: My goal here is to deliver this project as soon as possible, so I will only keep a certain set of features at very first. When we reach a stable and workable stage, I will start to evaluate different features and decide which of them to add. You know, there is priority to think about. For example, when you have A feature and B feature in the TODO list, A feature is awesome, you want to do it badly, however, your customers want B feature badly. I will implement B feature first.

There are many factors to think about the directions of a product, such as
- By adding this feature, how many users can benefit?
- By adding this feature, would anything be impacted?
- By adding this feature, how can I do the migration?
- How much it costs to deliver this feature?
- How many resource you have for this feature?
- When can I deliver this feature?

From my perspective, 80% of users would only need these kinds of interval. And only 20% or even lesser would need arbitrary intervals. For product sense, surely I will try my best to make 80% of users happy and try to make 20% users happy later.

Here I share some slides I wrote before









It is risky to bet valuable development time on a tons of features you think users would like it. However, the truth is, you would be wrong in most of cases. When users need something, they will tell you. When users need something badly, they will tell you again and again. So, I feel it is better to build only very core features, and see what users want, then satisfy them.

File: billy/models/__init__.py [hunk 5]
```diff
@@ -1,17 +1,36 @@
...
-from .payout.plan import PayoutPlan
-from .charge.plan import ChargePlan
-from .customers import Customer
-from .coupons import Coupon
-from .company import Company
+def setup_database(**settings):
+    """Setup database
+    
+    """
+    if 'engine' not in settings:
```
Comment (mahmoudimus) [inline billy/models/__init__.py:16]: We typically use `()` instead of `\` to denote multiline statements.

So:

``` python

settings['engine'] = (
      engine_from_config(settings, 'sqalchemy.')
)

```

Comment (mahmoudimus) [inline billy/models/__init__.py:16]: You should make this change going forward.

Comment (fangpenlin) [inline billy/models/__init__.py:16]: will do

Comment (fangpenlin) [inline billy/models/__init__.py:16]: done

**Comment by mahmoudimus** [comment]
@victorlin going to address one of your comments here:

You wrote -

> I got reasons to pass only GUID. Passing SQLALchemy table class instance around is error-prone in certain situations. For example, when you are dealing with _multi-threading_, passing object instance around causes problem. (emphasis added)

Since you're using a scoped_session and a sessionmaker, you shouldn't have to worry about multi-threading issues. 

> def process(records):
>     for record in records:
>         with transaction.manager:
>             record.num += 1
>             session.add(record)
> 
> pool = threading_pool()
> for thread in pool:
>     records = get_records(100)
>     thread.run(process, records)

This is a really inefficient way of doing bulk updates or insertions. If you want to do a bulk-update, using threading is typically a really poor way of doing it. There are constructs in SQLAlchemy that allow you to very easily do bulk updates or insertions. In one call, using one transaction.

> Let's say, we are processing transactions and we like to speed it up by using multi-threading or multi-processor. The code we wrote above looks fine, but it is not. You cannot create SQLALchemy object in one thread and pass it to another (exception will be raised). There is some thread-local stuff Session object is dealing with (sessions are bound to local threads). So, in this case, I prefer to make model only accept GUID argument rather than object.

There's other ways to do this if you want to pass a SQLAlchemy mapped object between threads. Simply, tell the ORM to stop tracking and instrumenting this object while you're passing it. Use `Session.expunge()`

Threads in Python are typically used for I/O heavy applications, so in that case, why not just pass the model, then do a `Session.merge()` and then do a `Session.refresh()`.

Passing in objects is a very nice way to work with relationships. Not using them really limits the power of SQLAlchemy.

File: billy/manage.py [hunk 6]
```diff
@@ -14,6 +14,7 @@ def create_tables():
     """
     Creates the tables if they dont exists
     """
+    print '#'*10, 'Base.metadata', Base.metadata
     Base.metadata.create_all(DB_ENGINE)
```
Comment (mahmoudimus) [inline billy/manage.py]: We should be using alembic for our database migrations.

File: billy/models/company.py [hunk 7]
```diff
@@ -1,179 +1,69 @@
...
+        """Update a company
 
-    @property
-    def processor(self):
         """
-        Get an instantiated processor for the company i.e DummyProcessor
-        :return: An instantiated ProcessorClass
+        company = self.get(guid, raise_error=True)
+        now = tables.now_func()
+        company.updated_at = now
```
Comment (mahmoudimus) [inline billy/models/company.py:68]: why not use `func.clock_timestamp()`

Comment (mahmoudimus) [inline billy/models/company.py:68]: oh that's right you're using sqlite, so why not `func.now()` ?

File: billy/models/tables.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    name = Column(Unicode(128))
+    #: is this company deleted?
+    deleted = Column(Boolean, default=False, nullable=False)
+    #: the created datetime of this company
+    created_at = Column(DateTime(timezone=True), default=now_func)
+    #: the updated datetime of this company
+    updated_at = Column(DateTime(timezone=True), default=now_func)
+
+    #: plans of this company
+    plans = relation('Plan', cascade='all, delete-orphan', backref='company')
```
Comment (mahmoudimus) [inline billy/models/tables.py]: `relation` is deprecated. use `relationship`.

**Comment by fangpenlin** [comment]
@mahmoudimus 

What I mean is not actually doing updates or insertions with threading. But for more complex business logic. For example, currently, we have `yield_transactions` method of `SubscriptionModel`. It is for yielding transactions from subscriptions. If we have a tons of subscription to process, it would be slow to do it in single thread. This kind of logic cannot be done with plain SQL expressions like bulk updates or insertions (it may could be, but might be too complex, or that needs to be optimized later).

I know there is a way I can leverage session to deals with cross thread issue. However, I prefer to keep things simple and stupid (KISS). The Session.refresh() will actually get the record from database by its primary key, just like what you do by passing GUID. So, in this case, for simplicity and consistency, l prefer passing primary keys. However, if passing object is more comfortable for youguys, I will refactory the model to do it the way later (When we reach the stable stage).

File: billy/models/transaction.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        payment_uri,
+        scheduled_at,
+    ):
+        """Create a transaction and return its ID
+
+        """
+        if transaction_type not in self.TYPE_ALL:
+            raise ValueError('Invalid transaction_type {}'.format(transaction_type))
+        transaction = tables.Transaction(
+            guid='TX' + make_guid(),
```
Comment (mahmoudimus) [inline billy/models/transaction.py:111]: This pattern should be moved to a base model.

File: billy/models/company.py (hunk 7 above)
Comment (fangpenlin) [inline billy/models/company.py:68]: Yeah, I am thinking about to use the timestamp from database as we may have multiple web nodes but single database. In that case we would like to have a consistent time over all nodes. However, by using the time stamp from database, it makes that a little bit hard to do unit tests. I will look at that and think about it later.

But for that `now_func`, it is actually a placeholder. As you can see the code

``` python

#: The now function for database relative operation
_now_func = [func.utc_timestamp]


def set_now_func(func):
    """Replace now function and return the old function

    """
    old = _now_func[0]
    _now_func[0] = func
    return old


def get_now_func():
    """Return current now func

    """
    return _now_func[0]


def now_func():
    """Return current datetime

    """
    func = _now_func[0]
    return func()
```

We can always replace that now function with anything you like. Surely you can replace it with a timestamp like

``` python
set_now_func(func.clock_timestamp)
```

File: billy/models/transaction.py (hunk 9 above)
Comment (fangpenlin) [inline billy/models/transaction.py:111]: I know we have duplicated code there. However, I will keep them there for a while. I would like to see what are those common elements shared by all modes, and extract them to a base class or something later. Sometimes by doing OO design too early, it would be over engineering or modeling it wrong. A little deferred refactorying would be helpful.

Comment (mahmoudimus) [inline billy/models/transaction.py:111]: We use this same pattern. I can share it here.

Comment (mahmoudimus) [inline billy/models/transaction.py:111]: @victorlin - definitely agree with you regarding OO design too early making it over engineering or modeling it wrong, but we've been working on this for almost 3 years now. We've got a couple of patterns that we can share w/ you that we've really nailed down over time.

One flaw I see with your method is if you ever have to override any of the uuid creation schemes, it's very hard to do so (you don't expose a method anywhere to do this).

I'll fork your branch of billy and make a pull request so I can demonstrate what I'm talking about. I'll also add you to our internal Balanced project so you can see how advanced it is.

File: billy/models/tables.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+            'plan.guid', 
+            ondelete='CASCADE', onupdate='CASCADE'
+        ), 
+        index=True,
+        nullable=False,
+    )
+    #: the discount of this subscription, 
+    #  e.g. 0.3 means 30% price off disscount
+    # TODO: maybe we should use decimal here? what about accuracy issue?
+    discount = Column(Float)
```
Comment (mahmoudimus) [inline billy/models/tables.py]: Nope. Never use float.

Comment (mahmoudimus) [inline billy/models/tables.py]: Use Numeric.

File: billy/models/transaction.py (hunk 9 above)
Comment (fangpenlin) [inline billy/models/transaction.py:111]: You're right. I am actually building billy with my experience and my practices. I have done things with this kind of structure and patterns for tons of projects. Although, I am doing this for years, I am still improving it gradually. 

The reason we have so many discussions here is that we have very different development experience and practices, surely it would take some time to get used to. And yes it would be nice if I can see how you do in the backend systems. I am sure I can learn something from that and apply on billy.

But for now, I can do it with the best productivity in my favor manners (in this week, maybe). So, I will keep working, and in the mean time, I will try get familiar with your style, then gradually refactory and transform into Balanced style. With good testing coverage, we can always switch between different implementation approaches easily. Does it make sense?

File: billy/models/plan.py (hunk 2 above)
Comment (remear) [inline billy/models/plan.py:72]: Do these intervals allow for subscriptions on a bi-weekly, bi-monthly, etc basis? This is already among the top requests I see in IRC when asked if Balanced has recurring billing capability.

File: billy/models/transaction.py (hunk 9 above)
Comment (mahmoudimus) [inline billy/models/transaction.py:111]: Absolutely :)

**Comment by mahmoudimus** [comment]
Also, are you going to recurring payout support? This is probably the most important distinguishing feature for Billy. See https://github.com/balanced/billy/issues/15.

File: billy/models/plan.py (hunk 2 above)
Comment (fangpenlin) [inline billy/models/plan.py:72]: @remear 
Yeah, I am thinking about multiple period interval. As they are needed, I will add them into this development cycle. 

There might could be a interval for subscription method like this

``` python
# charge every 3 months
billy.subscribe(
    customer_guid='XXX',
    plan_guid='XXX (monthly)',
    interval=3,
)
```

@victorlin TODO: add scenario for these cases to #25

**Comment by fangpenlin** [comment]
@mahmoudimus 

> Also, are you going to recurring payout support? This is probably the most important distinguishing feature for Billy. See #15.

Yes, recurring payout is included in this initial development cycle.

File: billy/models/tables.py (hunk 10 above)
Comment (fangpenlin) [inline billy/models/tables.py]: Will do.

File: billy/models/plan.py (hunk 2 above)
Comment (pnegahdar) [inline billy/models/plan.py:72]: Cool. Then perhaps mapping them to integers (limited sqlite data structures?) might reduce your flexibility and force you into a migration/interface modification later on, thats fine but could be avoided with a better representation.

File: billy/models/tables.py (hunk 8 above)
Comment (fangpenlin) [inline billy/models/tables.py]: done.

File: billy/models/tables.py (hunk 10 above)
Comment (fangpenlin) [inline billy/models/tables.py]: done.

**Comment by fangpenlin** [comment]
The features I am going to implement in this cycle are posted in #25 as usage scenario form. If you see anything that is needed badly but not there, please added it in a reply.

**Comment by fangpenlin** [comment]
Okay, for now, there are still details to take care, however, we have reached a rough workable version with 100% testing coverage. 

```
Name                                        Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------
billy                                          12      0   100%
billy.api                                       7      0   100%
billy.api.auth                                 36      0   100%
billy.api.company                               4      0   100%
billy.api.company.views                        25      0   100%
billy.api.customer                              4      0   100%
billy.api.customer.views                       27      0   100%
billy.api.plan                                  4      0   100%
billy.api.plan.views                           33      0   100%
billy.api.subscription                          4      0   100%
billy.api.subscription.views                   37      0   100%
billy.api.transaction                           3      0   100%
billy.api.transaction.views                    16      0   100%
billy.models                                   15      0   100%
billy.models.company                           42      0   100%
billy.models.customer                          36      0   100%
billy.models.plan                              50      0   100%
billy.models.processors                         0      0   100%
billy.models.processors.balanced_payments      63      0   100%
billy.models.processors.base                   10      0   100%
billy.models.schedule                          20      0   100%
billy.models.subscription                      97      0   100%
billy.models.tables                            91      0   100%
billy.models.transaction                       91      0   100%
billy.renderers                                31      0   100%
billy.request                                   7      0   100%
billy.scripts                                   0      0   100%
billy.scripts.initializedb                     19      0   100%
billy.scripts.process_transactions             32      0   100%
billy.utils                                     0      0   100%
billy.utils.generic                            26      0   100%
-------------------------------------------------------------------------
TOTAL                                         842      0   100%
----------------------------------------------------------------------
Ran 136 tests in 7.081s

OK
```

### To run the test

```
virtualenv --no-site-packages env
source env/bin/activate
python setup.py develop
pip install -r test_requirements.txt
python setup.py nosetests
# or you prefer 
nosetests
```

### Create tables database

```
initialize_billy_db development.ini
```

### Start the server

```
pserve development.ini --reload
```

### Create a company

```
# create company
curl http://127.0.0.1:6543/v1/companies/ -X POST \
    -d "processor_key=YOUR_BALANCED_API_KEY"
```

result

```
{
    "guid": "CPPEfEKswV4eywR67He2ZkFu", 
    "api_key": "DCwfL7zJi48EesUCLSiyemCuxxdSiB7T1Bha9icTBN4d", 
    "created_at": "2013-08-23T06:39:04.427220", 
    "updated_at": "2013-08-23T06:39:04.427232"
}
```

### Create a customer

```
# create customer
curl http://127.0.0.1:6543/v1/customers/ \
    -u DCwfL7zJi48EesUCLSiyemCuxxdSiB7T1Bha9icTBN4d: \
    -d "external_id=/v1/customers/AC1jqOF9TocQXGIXjuMVrpMu"
```

result

```
{
    "guid": "CUJn4Us9HqQmkEazeKDTsoYo", 
    "company_guid": "CPPEfEKswV4eywR67He2ZkFu", 
    "external_id": "/v1/customers/AC1jqOF9TocQXGIXjuMVrpMu", 
    "created_at": "2013-08-23T08:18:16.782940", 
    "updated_at": "2013-08-23T08:18:16.782950"
}
```

### Create a plan

```
# create plan
curl http://127.0.0.1:6543/v1/plans/ \
    -u DCwfL7zJi48EesUCLSiyemCuxxdSiB7T1Bha9icTBN4d: \
    -d "plan_type=charge" \
    -d "amount=55.66" \
    -d "frequency=daily"
```

result

```
{
    "guid": "PLWZH4N9mB5wTLwEDTSgaPpf",
    "company_guid": "CPPEfEKswV4eywR67He2ZkFu", 
    "interval": 1, 
    "plan_type": "charge", 
    "amount": "55.66", 
    "frequency": "daily", 
    "created_at": "2013-08-23T08:20:56.820769", 
    "updated_at": "2013-08-23T08:20:56.820785"
}
```

### Create a subscription

```
# create subscription
curl http://127.0.0.1:6543/v1/subscriptions/ \
    -u DCwfL7zJi48EesUCLSiyemCuxxdSiB7T1Bha9icTBN4d: \
    -d "customer_guid=CUJn4Us9HqQmkEazeKDTsoYo" \
    -d "plan_guid=PLWZH4N9mB5wTLwEDTSgaPpf"
```

result

```
{
    "guid": "SUBSHs7KijcmEmuqUFPEkNiP",
    "customer_guid": "CUJn4Us9HqQmkEazeKDTsoYo", 
    "plan_guid": "PLWZH4N9mB5wTLwEDTSgaPpf", 
    "period": 0, 
    "canceled": false, 
    "amount": "None", 
    "started_at": "2013-08-23T08:23:46.534374", 
    "next_transaction_at": "2013-08-23T08:23:46.534374", 
    "created_at": "2013-08-23T08:23:46.536058", 
    "updated_at": "2013-08-23T08:23:46.536065"
}
```

### Process transactions

We have a subscription in the database, and it should yield a transaction right away. To process them, here you call

```
process_billy_tx development.ini
```

A crontab setting or similar scheduler can be used to process all transactions periodically.

### What's next?

There are still some design decisions to make, such as should we allow modifying subscription, should we process transaction immediately when a subscription is added, I will think about them. I may create issues for them, then we can talk about it and see what kind of design is better. 

Although the testing coverage is 100%, doesn't mean it is bug-free nor perfect, I am sure there are many edge cases aren't tested. 

And for testing, we have unit tests and functional tests, they are all on SQLite. I will create a Vagrant environment and writes some integration tests there with PostgreSQL.

In the mean time, I may do some refactory, but the priority here are still the tests. With good testing coverage (include edge cases), we can move faster.

**Comment by mjallday** [comment]
@victorlin you are amazing :)

**Comment by fangpenlin** [comment]
@mjallday 
Thank you :)

Well, I think this branch is good to merge back now. It is stable enough and also functional. I will create issues for remaining task later.

**Comment by mjallday** [comment]
this pull-request is big and old. can we merge it so we don't have to look at 10k line pull requests ever again?

**Comment by mahmoudimus** [comment]
this looks great  +1
